### PR TITLE
Feature/add detail page

### PR DESCRIPTION
**Pull request by GPretadoR**
Added RecipeDetailsViewController
Added recipe nutrition call
Added analyzed instructions call
Added more coordinators
Added object models
Added similar recipes request
Added instructions step viewcontroller
Updated coordinators

### CODE REVIEW

File: RakutenShop/BaseClasses/BaseViewController/BaseViewController.swift [hunk 1]
```diff
@@ -26,6 +26,7 @@ class BaseViewController: UIViewController {
     }
 
     func setupView() {
+        self.view.backgroundColor = .lightGray
```
Comment (Haykmnatsakanyan) [inline RakutenShop/BaseClasses/BaseViewController/BaseViewController.swift]: I suggest to remove self )))

File: RakutenShop/Context/Context.swift [hunk 2]
```diff
@@ -8,4 +8,6 @@
 
 class Context {
     let services = Services()
+    
+    var imagesBaseUrl: String?
```
Comment (Haykmnatsakanyan) [inline RakutenShop/Context/Context.swift:12]: why did you add it here?

File: RakutenShop/Context/Services.swift [hunk 3]
```diff
@@ -10,7 +10,14 @@ class Services {
     let networkServiceProvider = NetworkServiceProvider(baseURL: Environments.baseUrl(env: .dev))
     
     let searchRecipesService: SearchRecipesService
+    let recipeNutritionService: GetNutritionsService
+    let recipeInstructionService: GetAnalyzedInstructionsService
+    let getSimilarRecipes: GetSimilarRecipesService
+    
     init() {
         self.searchRecipesService = SearchRecipesService(searchRequest: networkServiceProvider)
```
Comment (Haykmnatsakanyan) [inline RakutenShop/Context/Services.swift]: I suggest to remove selfs

File: RakutenShop/Model/AnalyzedInstructionsResponse/Entity.swift [hunk 4]
```diff
@@ -0,0 +1,7 @@
+import Foundation
```
Comment (Haykmnatsakanyan) [inline RakutenShop/Model/AnalyzedInstructionsResponse/Entity.swift:3]: add empty line ))

File: RakutenShop/Model/AnalyzedInstructionsResponse/Steps.swift [hunk 5]
```diff
@@ -0,0 +1,25 @@
+
+import Foundation
+struct Steps: Codable {
+	let number: Int?
+	let step: String?
+	let ingredients: [Entity]?
+	let equipment: [Entity]?
+
+	enum CodingKeys: String, CodingKey {
```
Comment (Haykmnatsakanyan) [inline RakutenShop/Model/AnalyzedInstructionsResponse/Steps.swift]: why do we need this?

File: RakutenShop/Networking/NetworkRequest.swift [hunk 6]
```diff
@@ -26,7 +26,7 @@ extension NetworkRequest {
     var queryParameters: [String: String]? { nil }
     var bodyParameters: [String: String]? { nil }
     var customHeaders: [String: String]? { nil }
-    var authorizationStrategy: AuthorizationStrategy? { nil }
+    var authorizationStrategy: AuthorizationStrategy? { .token }
```
Comment (Haykmnatsakanyan) [inline RakutenShop/Networking/NetworkRequest.swift]: question if it has default value, why is it optional?
the same for timeoutInterval

File: RakutenShop/Networking/NetworkServiceProvider.swift [hunk 7]
```diff
@@ -142,14 +142,16 @@ class NetworkServiceProvider: NetworkServices {
...
-        //                   urlRequest.headers.add(.authorization(bearerToken: token))
-        //               }
-        //           }
-        if var queryParameters = request.queryParameters {
-            queryParameters.updateValue("c5033dfeabae490e9e1cfc32e727eca0", forKey: "apiKey")
+        if let authorizationStrategy = request.authorizationStrategy, authorizationStrategy == .token {
+            let encoding = URLEncoding(destination: .queryString,
+                                       arrayEncoding: .noBrackets,
+                                       boolEncoding: .numeric)
+            urlRequest = try encoding.encode(urlRequest, with: ["apiKey": "c5033dfeabae490e9e1cfc32e727eca0"])
```
Comment (Haykmnatsakanyan) [inline RakutenShop/Networking/NetworkServiceProvider.swift:149]: I will suggest to write apiKey some different place

File: RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewController.swift [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+    override func setupViewModel() {
+        guard let viewModel = viewModel else { return }
+        
+        viewModel.isLoading.producer.startWithValues { (showLoading) in
+            DispatchQueue.main.async {
+                showLoading ? SVProgressHUD.show() : SVProgressHUD.dismiss()
+            }
+        }        
+        
+        viewModel.errorMessage.producer.skip(first: 1).startWithValues { (errorMessage) in
```
Comment (Haykmnatsakanyan) [inline RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewController.swift:53]: I suppose we you can just start instead of startWithValues which will not take default or first value, I am not sure, but you can try it.

File: RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewController.swift [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+                self.updateNutritions(nutritionsObject: nutritionsObject)
+            }
+        })
+        
+//        viewModel.analyzedInstructions.producer.startWithValues({ (instructions) in
+//            print(instructions)
+//        })
+    
+        reactive.makeBindingTarget { (weakself, object) in
+            self.instructionsTableView.reloadData()
```
Comment (Haykmnatsakanyan) [inline RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewController.swift]: self -> weakself
instead of ojbect -> _ just underscore

the same for next one.

File: RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewController.swift [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+    
+    // MARK: - Actions -
+    
+    @objc func saveButtonTapped() {
+        
+    }
+    
+    // MARK: - Helper functions -
+    
+    func updateNutritions(nutritionsObject: NutritionsResponseObject) {
```
Comment (Haykmnatsakanyan) [inline RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewController.swift]: I suppose this is private function

File: RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewController.swift [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+            let carbs = nutritionsObject.carbs,
+            let proteins = nutritionsObject.protein,
+            let fats = nutritionsObject.fat else { return }
+        caloriesLabel.text = R.string.localizable.recipeDetailCaloriesLabelText(calories)
+        carbsLabel.text = R.string.localizable.recipeDetailCarbsLabelText(carbs)
+        fatsLabel.text = R.string.localizable.recipeDetailFatsLabelText(fats)
+        proteinsLabel.text = R.string.localizable.recipeDetailProteinsLabelText(proteins)
+    }
+    
+    func updateUI(recipeObject: RecipeObject) {
```
Comment (Haykmnatsakanyan) [inline RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewController.swift]: make it private if possible

File: RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewController.swift [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+        self.title = recipeObject.title
```
Comment (Haykmnatsakanyan) [inline RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewController.swift]: remove self

File: RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewController.swift [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+        let imageName = recipeObject.image != nil ? recipeObject.image : recipeObject.imageUrls?.first
+        if let imgUrl = imageName, let url = viewModel?.buildImageUrl(imageName: imgUrl) {
+            self.recipeImageView.af_setImage(withURL: url)
```
Comment (Haykmnatsakanyan) [inline RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewController.swift:108]: remove self.

File: RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewCoordinator.swift [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+
+    private let context: Context
+    private var viewController: RecipeDetailViewController?
+    var recipeDetailCoordinator: StepListViewCoordinator?
+    var recipeObject: RecipeObject?
+    
+    init(context: Context, coordinator: BaseCoordinator) {
+        self.context = context
+        super.init(coordinator: coordinator)
+        self.viewController = R.storyboard.main.recipeDetailViewController()
```
Comment (Haykmnatsakanyan) [inline RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewCoordinator.swift]: remove self

File: RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewCoordinator.swift [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+        guard let viewController = viewController else { return }
+        let viewModel = RecipeDetailViewViewModel(context: context, coordinatorDelegate: self)
+        viewController.viewModel = viewModel
+    }
+    
+    override func start() {
+        guard let viewController = viewController, let rootNavController = rootViewController as? UINavigationController else { return }
+        rootNavController.pushViewController(viewController, animated: true)
```
Comment (Haykmnatsakanyan) [inline RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewCoordinator.swift:31]: it would be greate to have push function which is organizing this in the base classes, look the example in the Ihor's project.

File: RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewCoordinator.swift [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+        guard let recipeObject = recipeObject else { return }
+        viewController.viewModel?.recipeObject = MutableProperty<RecipeObject>(recipeObject)
```
Comment (Haykmnatsakanyan) [inline RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewCoordinator.swift]: I will suggest to move this line in the init function.

File: RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewCoordinator.swift [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+        viewController.viewModel?.getInfo()
```
Comment (Haykmnatsakanyan) [inline RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewCoordinator.swift]: I will suggest to call this function in the init of viewModel.

File: RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewCoordinator.swift [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+extension RecipeDetailViewCoordinator: RecipeDetailViewCoordinatorDelegate {
+    func didTapSaveButton() {
+        
+    }
+    
+    func didTapInstructionTableViewCell(instructions: AnalyzedInstructionsResponseObject) {
+        recipeDetailCoordinator = StepListViewCoordinator(context: context, coordinator: self)
+        recipeDetailCoordinator?.instructions = instructions
```
Comment (Haykmnatsakanyan) [inline RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewCoordinator.swift:42]: I will suggest to pass this value via constructor.

File: RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewViewModel.swift [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+class RecipeDetailViewViewModel: BaseViewModel {
+
+    private let context: Context
+    weak var coordinatorDelegate: RecipeDetailViewCoordinatorDelegate?
+    
+    var recipeObject: MutableProperty<RecipeObject>?
+    
+    var nutritionsObject = MutableProperty<NutritionsResponseObject>(NutritionsResponseObject(calories: "", carbs: "", fat: "", protein: "", bad: [], good: []))
```
Comment (Haykmnatsakanyan) [inline RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewViewModel.swift]: var nutritionsObject = MutableProperty<NutritionsResponseObject?>(nil)
I believe this is more correct approach, as in case of very big responseObject with 20 fields, it would be problem for you.

File: RakutenShop/Scenes/RecipesView/RecipesViewController.swift [hunk 20]
```diff
@@ -53,19 +57,19 @@ class RecipesViewController: BaseViewController {
             }
         }
         
+        viewModel.errorMessage.producer.skip(first: 1).startWithValues { (errorMessage) in
+            DispatchQueue.main.async {
+                SVProgressHUD.showError(withStatus: errorMessage)
+            }
+        }
+        
         viewModel.recipeItems.producer.skip(first: 1).startWithValues { (recipes) in
```
Comment (Haykmnatsakanyan) [inline RakutenShop/Scenes/RecipesView/RecipesViewController.swift:66]: maybe start function instead of startWithValues will be better, just try I am not sure.

File: RakutenShop/Scenes/RecipesView/RecipesViewCoordinator.swift [hunk 21]
```diff
@@ -32,8 +34,9 @@ class RecipesViewCoordinator: BaseCoordinator {
 }
 
 extension RecipesViewCoordinator: RecipesViewCoordinatorDelegate {
-    func didSelectItem(at indexPath: IndexPath) {
-        
+    func didSelectRecipe(recipe: RecipeObject) {
+        recipeDetailCoordinator = RecipeDetailViewCoordinator(context: context, coordinator: self)
+        recipeDetailCoordinator?.recipeObject = recipe
```
Comment (Haykmnatsakanyan) [inline RakutenShop/Scenes/RecipesView/RecipesViewCoordinator.swift:39]: I will just to pass this info via init.

File: RakutenShop/Scenes/StepListView/StepListViewController.swift [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        stepListTableView.dataSource = self
+    }
+
+    override func setupViewModel() {
+        guard let viewModel = viewModel else { return }
+        reactive.makeBindingTarget { [unowned self] (weakself, object) in
+            self.stepListTableView.reloadData()
```
Comment (Haykmnatsakanyan) [inline RakutenShop/Scenes/StepListView/StepListViewController.swift]: self -> weakself

File: RakutenShop/Scenes/StepListView/StepListViewCoordinator.swift [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+        self.viewController = R.storyboard.main.stepListViewController()
+        guard let viewController = viewController else { return }
+        let viewModel = StepListViewViewModel(context: context)
+        viewController.viewModel = viewModel
+    }
+    
+    override func start() {
+        guard let viewController = viewController, let rootNavController = rootViewController as? UINavigationController else { return }
+        rootNavController.pushViewController(viewController, animated: true)
+        guard let instructions = instructions else { return }
```
Comment (Haykmnatsakanyan) [inline RakutenShop/Scenes/StepListView/StepListViewCoordinator.swift]: move this line in the init

File: RakutenShop/Context/Context.swift (hunk 2 above)
Comment (GPretadoR) [reply RakutenShop/Context/Context.swift:12]: we neet it here :D

File: RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewController.swift (hunk 8 above)
Comment (GPretadoR) [reply RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewController.swift:53]: not for this case as I observe value change and show error popup

File: RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewCoordinator.swift (hunk 15 above)
Comment (GPretadoR) [reply RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewCoordinator.swift:31]: I will take a look at this on later stages

File: RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewCoordinator.swift (hunk 17 above)
Comment (GPretadoR) [reply RakutenShop/Scenes/RecipeDetailView/RecipeDetailViewCoordinator.swift]: This line should be executed afte viewcontroller is shown, otherwise binding will not work
